cpp/src/gandiva/external_c_functions.cc

```diff
@@ -0,0 +1,94 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License
+
+#include "llvm/IR/Type.h"
+
+#include "gandiva/engine.h"
+#include "gandiva/exported_funcs.h"
+
+namespace gandiva {
+static arrow::Result<llvm::Type*> AsLLVMType(const DataTypePtr& from_type,
+                                             LLVMTypes* types) {
+  switch (from_type->id()) {
+    case arrow::Type::BOOL:
+      return types->i1_type();
+    case arrow::Type::INT8:
+    case arrow::Type::UINT8:
+      return types->i8_type();
+    case arrow::Type::INT16:
+    case arrow::Type::UINT16:
+      return types->i16_type();
+    case arrow::Type::INT32:
+    case arrow::Type::UINT32:
+      return types->i32_type();
+    case arrow::Type::INT64:
+    case arrow::Type::UINT64:
+      return types->i64_type();
+    case arrow::Type::FLOAT:
+      return types->float_type();
+    case arrow::Type::DOUBLE:
+      return types->double_type();
+    default:
+      return Status::NotImplemented("Unsupported arrow data type: " +
+                                    from_type->ToString());
+  }
+}
+
+// map from a NativeFunction's signature to the corresponding LLVM signature
+static arrow::Result<std::pair<std::vector<llvm::Type*>, llvm::Type*>> MapToLLVMSignature(
+    const FunctionSignature& sig, const NativeFunction& func, LLVMTypes* types) {
+  std::vector<llvm::Type*> arg_llvm_types;
+  arg_llvm_types.reserve(sig.param_types().size());
```

Comment:
Initially I don't want to calculate the size since it may duplicate some code but it doesn't seem too bad after I write it, and I add a new function to calculate the number of arguments up front now.